controllers/storageclusterpeer/storageclusterpeer_controller.go

```diff
@@ -18,46 +18,488 @@ package storageclusterpeer
+			}
+			return true
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			// Check if the annotation is not present
+			if e.Object.GetAnnotations() == nil {
+				return false
+			}
+			if _, exists := e.Object.GetAnnotations()[util.CephBlockPoolBlackListAnnotation]; !exists {
+				return false
+			}
+			return true
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			if e.ObjectOld == nil || e.ObjectNew == nil {
+				return false
+			}
+			oldObj := e.ObjectOld.(*rookCephv1.CephBlockPool)
+			newObj := e.ObjectNew.(*rookCephv1.CephBlockPool)
+
+			// requeue only if changes are made for
+			return !(reflect.DeepEqual(oldObj.Spec.Mirroring, newObj.Spec.Mirroring) ||
+				reflect.DeepEqual(oldObj.GetAnnotations(), newObj.GetAnnotations()))
+		},
+	}
+
+	secretPredicate := predicate.Funcs{
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return strings.HasPrefix(e.Object.GetName(), bootstrapSecretPrefix)
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return strings.HasPrefix(e.ObjectOld.GetName(), bootstrapSecretPrefix)
+		},
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1.StorageClusterPeer{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		Owns(&rookCephv1.CephBlockPool{}, builder.WithPredicates(cephBlockPoolPredicate)).
+		Owns(&corev1.Secret{}, builder.WithPredicates(secretPredicate)).
+		Owns(&rookCephv1.CephRBDMirror{}).
+		Complete(r)
 }
 
 //+kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclusterpeers,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclusterpeers/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclusterpeers/finalizers,verbs=update
+//+kubebuilder:rbac:groups=ceph.rook.io,resources=cephblockpools;,verbs=list;update;
+//+kubebuilder:rbac:groups=ceph.rook.io,resources=cephrbdmirrors;,verbs=create;update
+//+kubebuilder:rbac:groups=core,resources=secrets,verbs=create;update;delete
+//+kubebuilder:rbac:groups=config.openshift.io,resources=clusterversions,verbs=get
 
 // Reconcile is part of the main kubernetes reconciliation loop which aims to
 // move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the StorageClusterPeer object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.0/pkg/reconcile
-func (r *StorageClusterPeerReconciler) Reconcile(ctx context.Context, _ ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+func (r *StorageClusterPeerReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	var err error
+	r.ctx = ctrllog.IntoContext(ctx, r.Log)
+	r.Log.Info("Reconciling StorageClusterPeer.")
```

Comment:
in the logs, could you pls check in this line, how many times the context is being logged? I wasn't able to find the offending line and probably think supplied param `ctx` already has object and again `ctrllog.IntoContext` duplicates it.